extensions/consensus/consensus.go

```diff
@@ -0,0 +1,240 @@
+// aliases, that may modify the logic in certain well-defined ways. Arbitrary
+// changes to kwild consensus logic may require changes to kwild internals (as
+// is the standard approach in most blockchains), or adding new capabilities to
+// this consensus extensions package and how kwild uses it to facilitate new
+// types of live changes.
+package consensus
+
+import (
+	"context"
+	"fmt"
+	"slices"
+
+	"github.com/kwilteam/kwil-db/common"
+	"github.com/kwilteam/kwil-db/common/chain"
+	"github.com/kwilteam/kwil-db/core/types/serialize"
+	"github.com/kwilteam/kwil-db/core/types/transactions"
+	"github.com/kwilteam/kwil-db/extensions/resolutions"
+)
+
+// Need fork order defined for unambiguous application of many at same height,
+// such as is common with genesis on a new network.
+
+// Hardforks contains all known hardfork definitions.  Include both
+// canonical/kwild and extensions? This is just a list. When kwild loads
+// genesis.json, it locates the named hardforks in this registry.
+var Hardforks = map[string]*Hardfork{}
+
+// RegisterHardfork registers a hardfork definition by the name that should be
+// expected in the genesis file. The payload type is registered in the
+// transaction package, but the route is enabled in the tx app at activation.
+func RegisterHardfork(hf *Hardfork) {
+	if _, have := Hardforks[hf.Name]; have {
+		panic(fmt.Sprintf("already have hardfork %q", hf.Name))
+	}
+	Hardforks[hf.Name] = hf
+
+	for _, newPayload := range hf.TxPayloads {
+		transactions.RegisterPayload(newPayload.Type)
+		// NOTE: newPayload.Route stays in the []Hardfork.TxPayloads until
+		// activation, when TxApp/ABCI enable the tx route.
+	}
+}
+
+func HaveHardFork(name string) bool {
+	_, have := Hardforks[name]
+	return have
+}
+
+// Hardfork specifies the fundamental changes affected by a named hardfork. If
+// a field is nil or the zero value for the type, that particular change is not
+// part of the hardfork's definition. When the hardfork is registered, only the
+// defined changes are applied in the relevant parts of kwild.
+type Hardfork struct {
+	// Name is the hard fork's code name.
+	Name string
+	// NOTE: Activation height is specified by genesis.json (or other dynamic
+	// methods like signaling / voting). This struct defines the changes.
+
+	// Optional consensus logic overrides below.  All of these will require
+	// dynamic kwil-db support, such as registering new encoders and transaction
+	// payloads, submitting consensus parameter updates to the consensus engine,
+	// applying state adjustments at activation height in ABCI / TxApp, etc.
+
+	// TxPayloads specifies new transaction payload to recognize at activation.
+	// To modify an existing payload in a backward incompatible way, instead
+	// create a new version of the payload such as PayloadTypeExecuteActionV2.
+	// TODO: consider payload removal/invalidation and replace.
+	TxPayloads []Payload // Type() and tx app Route() implementation
```

Comment:
I'm not sure if we're discussing this in the same context, because this is to do just that.  It gives network operators (and the core kwil-db dev team) ways to add or potentially modify payloads with some level of coordination because validators cannot upgrade their software at the same instant in practice.  And removal of payloads (although not currently supported like with the resolution mods) necessitates an associated height otherwise we cannot replay any blocks or sync new nodes.

With `TxPayloads`, network X operators would define new playloads in the extension system, and have them either activated at genesis if they are planning a new network, or at some arbitrary point in their existing network's timeline, such as if they conceive of a need for a new payload type that they want to enable without breaking the network.

New payloads becoming available smoothly in a network upgrade to even the official kwil-db code base (canonical) necessitate this too.  But with an extended fork, that would mean they are too specific to network X to be part of the kwil-db official code base, and are thus done as an extended fork so hacking the code is not necessary.

This is just one way to handle new payloads.  Registering them at activation is one way to have the tx router only use them when it is time, but conditional logic could exist in the routes as well.